### ISSUE TITLE
idea IDE cannot resolve the generated classes

### ISSUE THREAD

**Issue by ikidou** [tbroyer/gradle-apt-plugin#32]
Hello Thomas,
When I use this plugin in Android Studio(idea based),source code and classes was generated,but the android studio can't resolve the generated classes and build successful. Any advice for me?

**Comment by tbroyer** [comment]
Did you review the doc https://github.com/tbroyer/gradle-apt-plugin#intellij-idea and previous issues related to IDEA: #1, https://github.com/tbroyer/gradle-apt-plugin/pull/4#issuecomment-221650196, #6, #26 ?

**Comment by ikidou** [comment]
I already apply 'idea' plugin to my `build.gradle`,but I got the source config from `Project Structure` only "build/generated/source/apt/test" folder added to `Test Source Folders`. Of course I tried upgrade gradle 3.0.

**Comment by ikidou** [comment]
Change `build/generated/source/apt/` to another location such as `generated/source/apt/`,then idea works OK. because "$project/build" is excluded by default.

**Comment by tbroyer** [comment]
This plugin is supposed to “un-exclude” the `${project.buildDir}` and re-exclude each of its subdirectories (with the exception of `build/generated`, declaring `build/generated/source/apt` as a “generated sources” directory); but it only properly works if the project had the `idea` plugin applied before you import it in Android Studio. Otherwise, after you build once and refresh the Gradle project, the `build/` folder won't be un-excluded (but all others will be properly declared), so you'll have to manually exclude it.
(just tested with Android Studio 2.1.3 on a small test project, tried many different scenarios)

BTW, I'm not going to change the default to a folder outside the `$project.buildDir`, because those sources are a “build artifact” (from Gradle's point of view) and this is where they're supposed to go. But it wouldn't even solve the problem; see below.

The problem here has more to do with Android Studio: for some reason (that only Google knows), Android Studio does not have the _Annotation Processors_ screen in a Gradle project's Settings (it's there in the Default Settings though), and seems to hard-code it to output generated sources to `build/classes/main/generated` (whereas classes are output to `build/classes/main`, which BTW means that your generated sources now become resources of your project and end up being bundled into the JAR if you don't pay attention and don't `./gradlew clean` before you `./gradlew assemble`). This also means that if you get Android Studio "properly" configured from the plugin, using “Make” in Android Studio will actually fail with a `java: duplicate class: test.AutoValue_Test` kind of error because the Java compiler will see the generated sources from the Gradle build in `build/generated/source/apt/main` and the ones from the Android Studio build in `build/classes/main/generated`).

**TL;DR: Android Studio has stupid defaults that don't even work properly**: annotation processing seems to work (i.e. “Make” in Android Studio succeeds) but the editor can't see the generated files (in `build/classes/main/generated`), plus you risk packaging them in your JAR (see above).
And if you `apply plugin: 'idea'` and trigger this plugin's configuration, you'll get "duplicate class" errors.
And if you configure your project to output generated files where Android Studio generates them:
```gradle
sourceSets.all {
    output.generatedSourcesDir = file("${buildDir}/classes/$name/generated")
}
```
it won't work much better (and then you're _guaranteed_ to bundle your generated sources into your final JAR).

**Comment by fab1an** [comment]
I wrestled with this and arrived at the conclusion:

1. Don't add the sourceset explicitely for the reasons above and because AndroidStudio will then show them mixed up with the normal sources in the packages-view.

2. Just, whenever you synchronize with gradle, open the `<module>.iml` file afterwards and delete the line `<excludeFolder url="file://$MODULE_DIR$/build" />`

3. Click `File -> Synchronise`, after that AndroidStudio will see your sources.
